libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,159 @@
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/ControlFlowRevertPruner.h>
+
+#include <libsolutil/Algorithms.h>
+
+namespace solidity::frontend
+{
+
+void ControlFlowRevertPruner::run()
+{
+	for (auto& [pair, flow]: m_cfg.allFunctionFlows())
+		collectCalls(*pair.function, pair.contract);
+
+	// In the second iteration we process anything left pending
+	removeRevertingNodes();
+}
+
+FunctionDefinition const* ControlFlowRevertPruner::resolveCall(FunctionCall const& _functionCall, ContractDefinition const* _contract)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return nullptr;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+	{
+		if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+		{
+			if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+				if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+				{
+					solAssert(contractType->isSuper(), "");
+					ContractDefinition const* superContract = contractType->contractDefinition().superContract(*_contract);
+
+					return &unresolvedFunctionDefinition.resolveVirtual(
+						*_contract,
+						superContract
+					);
+				}
+		}
+		else
+		{
+			solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+			return &unresolvedFunctionDefinition;
+		}
+	}
+	else if (auto const* identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+	{
+		solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+		return &unresolvedFunctionDefinition.resolveVirtual(*_contract);
+	}
+
+	return &unresolvedFunctionDefinition;
+}
+
+void ControlFlowRevertPruner::removeRevertingNodes()
+{
+	std::set<CFG::FunctionContractTuple> pending_functions = keys(m_functions);
+
+	while (!pending_functions.empty())
+	{
+		CFG::FunctionContractTuple item = *pending_functions.begin();
+		pending_functions.erase(pending_functions.begin());
+
+		auto& revertState = m_functions[item];
+
+		auto previousState = revertState;
+		revertState = RevertState::AllPathsRevert;
+
+		FunctionFlow const& functionFlow = m_cfg.functionFlow(*item.function, item.contract);
+
+		solidity::util::BreadthFirstSearch<CFGNode*>{{functionFlow.entry}}.run(
+			[&](CFGNode* _node, auto&& _addChild) {
+				if (_node == functionFlow.exit)
```

Comment:
Can it happen that the node is an exit node but still has reverting function calls? Can you add an assertion that in this case, there are no function calls?

Reply:
I don't think this can happen. The exit node is created when the function flow is created and exits are created by just connecting to the one exit node.